File: src/main/java/io/dddrive/ddd/model/base/AggregateRepositoryBase.java

package io.dddrive.ddd.model.base;

import static io.dddrive.util.Invariant.assertThis;
import static io.dddrive.util.Invariant.requireThis;

import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.springframework.context.ApplicationEvent;

import io.crnk.core.queryspec.FilterOperator;
import io.crnk.core.queryspec.PathSpec;
import io.crnk.core.queryspec.QuerySpec;
import io.dddrive.app.event.AggregateStoredEvent;
import io.dddrive.app.service.api.AppContext;
import io.dddrive.app.service.api.AppContextSPI;
import io.dddrive.ddd.model.Aggregate;
import io.dddrive.ddd.model.AggregatePersistenceProvider;
import io.dddrive.ddd.model.AggregateRepository;
import io.dddrive.ddd.model.PartRepository;
import io.dddrive.ddd.model.enums.CodeAggregateType;
import io.dddrive.ddd.model.enums.CodeAggregateTypeEnum;
import io.dddrive.property.model.PropertyProvider;
import io.dddrive.property.model.impl.PropertyFilter;
import io.dddrive.property.model.impl.PropertyHandler;
import io.dddrive.search.service.api.SearchService;
import javassist.util.proxy.ProxyFactory;

public abstract class AggregateRepositoryBase<A extends Aggregate, V extends Object>
		implements AggregateRepository<A, V>, AggregateRepositorySPI<A, V> {

	private final Class<? extends Aggregate> intfClass;
	private final String aggregateTypeId;
	private final AppContext appContext;

	private final List<PartRepository<? super A, ?>> partRepositories = new ArrayList<>();

	private final ProxyFactory proxyFactory;
	private final Class<?>[] proxyFactoryParamTypeList;

	private boolean didAfterCreate = false;
	private boolean didAfterLoad = false;
	private boolean didBeforeStore = false;
	private boolean didAfterStore = false;

	protected AggregateRepositoryBase(
			Class<? extends AggregateRepository<A, V>> repoIntfClass,
			Class<? extends Aggregate> intfClass,
			Class<? extends Aggregate> baseClass,
			String aggregateTypeId,
			AppContext appContext) {
		this.intfClass = intfClass;
		this.aggregateTypeId = aggregateTypeId;
		this.appContext = appContext;
		this.proxyFactory = new ProxyFactory();
		this.proxyFactory.setSuperclass(baseClass);
		this.proxyFactory.setFilter(PropertyFilter.INSTANCE);
		this.proxyFactoryParamTypeList = new Class<?>[] { repoIntfClass, Object.class };
		((AppContextSPI) appContext).addRepository(intfClass, this);
	}

	@Override
	@SuppressWarnings("unchecked")
	public final AggregatePersistenceProvider<A> getPersistenceProvider() {
		return (AggregatePersistenceProvider<A>) this.appContext.getAggregatePersistenceProvider(this.intfClass);
	}

	@Override
	public final PropertyProvider getPropertyProvider() {
		return this.appContext.getPropertyProvider(this.intfClass);
	}

	@Override
	@SuppressWarnings("unchecked")
	public final Class<A> getAggregateClass() {
		return (Class<A>) this.intfClass;
	}

	@Override
	public final CodeAggregateType getAggregateType() {
		return CodeAggregateTypeEnum.getAggregateType(this.aggregateTypeId);
	}

	@Override
	public final AppContext getAppContext() {
		return this.appContext;
	}

	protected void addPartRepository(PartRepository<? super A, ?> partRepository) {
		requireThis(partRepository != null, "partRepository is not null");
		this.partRepositories.add(partRepository);
	}

	public boolean hasAccountId() {
		return true;
	}

	@Override
	@SuppressWarnings("unchecked")
	public final A newAggregate(Object state) {
		A aggregate = null;
		try {
			Object[] params = new Object[] { this, state };
			aggregate = (A) this.proxyFactory.create(this.proxyFactoryParamTypeList, params, PropertyHandler.INSTANCE);
		} catch (NoSuchMethodException | IllegalArgumentException | InstantiationException | IllegalAccessException
				| InvocationTargetException e) {
			e.printStackTrace();
			throw new RuntimeException(this.getClassName() + ": could not create aggregate");
		}
		return aggregate;
	}

	@Override
	public final A create(Integer tenantId) {

		AggregatePersistenceProvider<A> persistenceProvider = this.getPersistenceProvider();
		Integer aggregateId = persistenceProvider.nextAggregateId();
		A aggregate = persistenceProvider.doCreate();

		Integer doInitSeqNr = ((AggregateBase) aggregate).doInitSeqNr;
		((AggregateSPI) aggregate).doInit(aggregateId, tenantId);
		assertThis(((AggregateBase) aggregate).doInitSeqNr > doInitSeqNr,
				this.getClassName(aggregate) + ": doInit was propagated");

		this.doInitParts(aggregate);

		aggregate.calcAll();

		this.didAfterCreate = false;
		this.doAfterCreate(aggregate);
		assertThis(this.didAfterCreate, this.getClassName() + ": doAfterCreate was propagated");

		return aggregate;
	}

	@Override
	public final void doInitParts(A aggregate) {
		for (PartRepository<? super A, ?> partRepo : this.partRepositories) {
			partRepo.init(aggregate);
		}
	}

	@Override
	public void doAfterCreate(A aggregate) {
		this.didAfterCreate = true;
		Integer doAfterCreateSeqNr = ((AggregateBase) aggregate).doAfterCreateSeqNr;
		((AggregateSPI) aggregate).doAfterCreate();
		assertThis(((AggregateBase) aggregate).doAfterCreateSeqNr > doAfterCreateSeqNr,
				this.getClassName(aggregate) + ": doAfterCreate was propagated");
	}

	@Override
	public final A get(Integer id) {

		requireThis(id != null, "id not null");
		AggregatePersistenceProvider<A> persistenceProvider = this.getPersistenceProvider();
		A aggregate = persistenceProvider.doLoad(id);

		this.doInitParts(aggregate);
		this.doLoadParts(aggregate);

		Integer doAssignPartsSeqNr = ((AggregateBase) aggregate).doAssignPartsSeqNr;
		((AggregateSPI) aggregate).doAssignParts();
		assertThis(((AggregateBase) aggregate).doAssignPartsSeqNr > doAssignPartsSeqNr,
				this.getClassName(aggregate) + ": doAssignParts was propagated");

		aggregate.calcVolatile();

		this.didAfterLoad = false;
		this.doAfterLoad(aggregate);
		assertThis(this.didAfterLoad, this.getClassName() + ": doAfterLoad was propagated");

		return aggregate;
	}

	@Override
	public final void doLoadParts(A aggregate) {
		List<PartRepository<? super A, ?>> repos = new ArrayList<>(this.partRepositories);
		Collections.reverse(repos);
		for (PartRepository<? super A, ?> partRepo : repos) {
			partRepo.load(aggregate);
		}
	}

	@Override
	public void doAfterLoad(A aggregate) {
		this.didAfterLoad = true;
		Integer doAfterLoadSeqNr = ((AggregateBase) aggregate).doAfterLoadSeqNr;
		((AggregateSPI) aggregate).doAfterLoad();
		assertThis(((AggregateBase) aggregate).doAfterLoadSeqNr > doAfterLoadSeqNr,
				this.getClassName(aggregate) + ": doAfterLoad was propagated");
	}

	@Override
	public final void discard(A aggregate) {
		((AggregateBase) aggregate).setStale();
	}

	@Override
	public final void store(A aggregate) {

		this.didBeforeStore = false;
		this.doBeforeStore(aggregate);
		assertThis(this.didBeforeStore, this.getClassName() + ": doBeforeStore was propagated");

		this.getPersistenceProvider().doStore(aggregate);
		this.doStoreParts(aggregate);
		this.storeSearch(aggregate);

		this.didAfterStore = false;
		this.doAfterStore(aggregate);
		assertThis(this.didAfterStore, this.getClassName() + ": doAfterStore was propagated");

	}

	@Override
	public void doBeforeStore(A aggregate) {
		this.didBeforeStore = true;
		Integer doBeforeStoreSeqNr = ((AggregateBase) aggregate).doBeforeStoreSeqNr;
		((AggregateSPI) aggregate).doBeforeStore();
		assertThis(((AggregateBase) aggregate).doBeforeStoreSeqNr > doBeforeStoreSeqNr,
				this.getClassName(aggregate) + ": doBeforeStore was propagated");
	}

	@Override
	public final void doStoreParts(A aggregate) { // TODO: take away from interface
		for (PartRepository<? super A, ?> partRepo : this.partRepositories) {
			partRepo.store(aggregate);
		}
	}

	private final void storeSearch(A aggregate) {
		((AggregateBase) aggregate).doStoreSearch();
	}

	protected final void storeSearch(Aggregate aggregate, List<String> texts, List<String> tokens) {
		this.appContext.getBean(SearchService.class).storeSearch(aggregate, texts, tokens);
	}

	@Override
	public void doAfterStore(A aggregate) {
		this.didAfterStore = true;
		Integer doAfterStoreSeqNr = ((AggregateBase) aggregate).doAfterStoreSeqNr;
		((AggregateSPI) aggregate).doAfterStore();
		assertThis(((AggregateBase) aggregate).doAfterStoreSeqNr > doAfterStoreSeqNr,
				this.getClassName(aggregate) + ": doAfterStore was propagated");
		this.discard(aggregate);
		ApplicationEvent aggregateStoredEvent = new AggregateStoredEvent(aggregate, aggregate);
		this.appContext.publishApplicationEvent(aggregateStoredEvent);
	}

	@Override
	public abstract List<V> doFind(QuerySpec querySpec);

	@Override
	public final List<V> getByForeignKey(String fkName, Integer targetId) {
		QuerySpec querySpec = new QuerySpec(Aggregate.class);
		querySpec.addFilter(PathSpec.of(fkName).filter(FilterOperator.EQ, targetId));
		return this.find(querySpec);
	}

	protected String getClassName() {
		return this.getClass().getSimpleName();
	}

	protected String getClassName(A aggregate) {
		return aggregate.getClass().getSuperclass().getSimpleName();
	}

}

File: fm-dddrive/src/main/java/dddrive/ddd/core/model/base/AggregateRepositoryBase.kt
package dddrive.ddd.core.model.base

import com.github.benmanes.caffeine.cache.Caffeine
import dddrive.ddd.core.model.Aggregate
import dddrive.ddd.core.model.AggregatePersistenceProvider
import dddrive.ddd.core.model.AggregateRepository
import dddrive.ddd.core.model.AggregateRepositorySPI
import dddrive.ddd.core.model.AggregateSPI
import dddrive.ddd.core.model.Part
import dddrive.ddd.core.model.PartRepository
import dddrive.ddd.core.model.RepositoryDirectory
import dddrive.ddd.core.model.RepositoryDirectory.Companion.instance
import dddrive.ddd.core.model.RepositoryDirectorySPI
import dddrive.ddd.core.model.enums.CodeAggregateType
import dddrive.ddd.core.model.enums.CodeAggregateTypeEnum
import dddrive.ddd.core.model.impl.PartRepositoryImpl
import dddrive.ddd.property.model.Property
import java.time.OffsetDateTime
import java.util.function.Function

abstract class AggregateRepositoryBase<A : Aggregate>(
	private val intfClass: Class<out Aggregate>,
	private val aggregateTypeId: String,
) : AggregateRepository<A>,
	AggregateRepositorySPI<A> {

	private val objCache = Caffeine
		.newBuilder()
		.maximumSize(200)
		.recordStats()
		.build<Any, A>()

	private var _persistenceProvider: AggregatePersistenceProvider<A>? = null
	private var didAfterCreate = false
	private var didAfterLoad = false
	private var didBeforeStore = false
	private var didAfterStore = false

	init {
		(this.directory as RepositoryDirectorySPI).addRepository(intfClass, this)
		this.registerParts()
	}

	override val aggregateType: CodeAggregateType
		get() = CodeAggregateTypeEnum.getAggregateType(this.aggregateTypeId)

	override val directory: RepositoryDirectory
		get() = instance

	@Suppress("UNCHECKED_CAST")
	// final TODO
	override val persistenceProvider: AggregatePersistenceProvider<A>
		get() {
			if (_persistenceProvider == null) {
				_persistenceProvider =
					directory.getPersistenceProvider(intfClass) as AggregatePersistenceProvider<A>
			}
			return _persistenceProvider!!
		}

	override fun idToString(id: Any?): String? = if (id == null) null else this.persistenceProvider.idToString(id)

	override fun idFromString(id: String?): Any? = if (id == null) null else this.persistenceProvider.idFromString(id)

	override fun doLogChange(property: String): Boolean = !NotLoggedProperties.contains(property)

	protected fun <AA : Aggregate, PP : Part<AA>> addPart(
		partIntfClass: Class<out PP>,
		factory: (AA, PartRepository<AA, PP>, Property<*>, Int) -> PP,
	) {
		val partRepository: PartRepository<AA, PP> = PartRepositoryImpl(partIntfClass, factory)
		(this.directory as RepositoryDirectorySPI).addPartRepository(partIntfClass, partRepository)
	}

	override fun create(
		tenantId: Any,
		userId: Any,
		timestamp: OffsetDateTime,
	): A {
		val persistenceProvider = this.persistenceProvider
		val aggregateId = persistenceProvider.nextAggregateId()
		val aggregate = this.createAggregate(true)

		try {
			aggregate.meta.disableCalc()
			val doCreateSeqNr = (aggregate as AggregateBase).doCreateSeqNr
			aggregate.doCreate(aggregateId, tenantId)
			check(aggregate.doCreateSeqNr > doCreateSeqNr) {
				intfClass.simpleName + ": doCreate was propagated"
			}
		} finally {
			aggregate.meta.enableCalc()
		}

		aggregate.meta.calcAll() // TODO reconsider, currently must be here because docs are frozen in
		// doAfterCreate

		this.didAfterCreate = false
		this.doAfterCreate(aggregate, userId, timestamp)
		check(this.didAfterCreate) { intfClass.simpleName + ": doAfterCreate was propagated" }

		return aggregate
	}

	override fun doAfterCreate(
		aggregate: A,
		userId: Any,
		timestamp: OffsetDateTime,
	) {
		this.didAfterCreate = true
		val doAfterCreateSeqNr = (aggregate as AggregateBase).doAfterCreateSeqNr
		(aggregate as AggregateSPI).doAfterCreate(userId, timestamp)
		check(aggregate.doAfterCreateSeqNr > doAfterCreateSeqNr) {
			intfClass.simpleName + ": doAfterCreate was propagated"
		}
	}

	override fun get(id: Any): A = this.objCache.get(id, Function { aggrId: Any? -> this.get(aggrId!!, true) })

	override fun load(id: Any): A = this.get(id, false)

	private fun get(
		id: Any,
		isFrozen: Boolean,
	): A {
		val persistenceProvider = this.persistenceProvider
		check(persistenceProvider.isValidId(id)) {
			"valid id " + id + " (" + id.javaClass.getSimpleName() + ")"
		}

		val aggregate = this.createAggregate(false)

		(aggregate as AggregateSPI).beginLoad()
		persistenceProvider.doLoad(aggregate, id)
		aggregate.endLoad()

		aggregate.meta.calcVolatile()

		if (isFrozen) {
			(aggregate as AggregateBase).freeze()
		}

		this.didAfterLoad = false
		this.doAfterLoad(aggregate)
		check(this.didAfterLoad) { intfClass.simpleName + ": doAfterLoad was propagated" }

		return aggregate
	}

	/**
	 * Create a new aggregate instance. Concrete repositories must override this to directly
	 * instantiate their Impl class.
	 */
	protected abstract fun createAggregate(isNew: Boolean): A

	override fun doAfterLoad(aggregate: A) {
		this.didAfterLoad = true
		val doAfterLoadSeqNr = (aggregate as AggregateBase).doAfterLoadSeqNr
		(aggregate as AggregateSPI).doAfterLoad()
		check(aggregate.doAfterLoadSeqNr > doAfterLoadSeqNr) {
			intfClass.simpleName + ": doAfterLoad was propagated"
		}
	}

	override fun store(
		aggregate: A,
		userId: Any,
		timestamp: OffsetDateTime,
	) {
		try {
			this.didBeforeStore = false
			this.doBeforeStore(aggregate, userId, timestamp)
			check(this.didBeforeStore) { intfClass.simpleName + ": doBeforeStore was propagated" }

			this.persistenceProvider.doStore(aggregate)

			this.didAfterStore = false
			this.doAfterStore(aggregate)
			check(this.didAfterStore) { intfClass.simpleName + ": doAfterStore was propagated" }
		} catch (e: Exception) {
			throw RuntimeException("${intfClass.simpleName}: could not store aggregate (${e.message})", e)
		}
	}

	override fun doBeforeStore(
		aggregate: A,
		userId: Any,
		timestamp: OffsetDateTime,
	) {
		this.didBeforeStore = true
		val doBeforeStoreSeqNr = (aggregate as AggregateBase).doBeforeStoreSeqNr
		(aggregate as AggregateSPI).doBeforeStore(userId, timestamp)
		check(aggregate.doBeforeStoreSeqNr > doBeforeStoreSeqNr) {
			intfClass.simpleName + ": doBeforeStore was propagated"
		}
	}

	override fun doAfterStore(aggregate: A) {
		this.didAfterStore = true
		val doAfterStoreSeqNr = (aggregate as AggregateBase).doAfterStoreSeqNr
		(aggregate as AggregateSPI).doAfterStore()
		check(aggregate.doAfterStoreSeqNr > doAfterStoreSeqNr) {
			intfClass.simpleName + ": doAfterStore was propagated"
		}
		this.handleAggregateStored(aggregate.id)
	}

	override fun getByForeignKey(
		fkName: String,
		targetId: Any,
	): List<Any> = this.persistenceProvider.getByForeignKey(fkName, targetId)

	fun handleAggregateStored(id: Any) {
		if (this.objCache.getIfPresent(id) != null) {
			this.objCache.invalidate(id)
		}
	}

	companion object {

		private val NotLoggedProperties =
			setOf(
				"id",
				"maxPartId",
				"version",
				"createdByUser",
				"createdAt",
				"modifiedByUser",
				"modifiedAt",
			)
	}

}
